src/main/kotlin/com/cultureamp/eventsourcing/PGSessionAdvisoryBookmarkLock.kt

```diff
@@ -0,0 +1,70 @@
+package com.cultureamp.eventsourcing
+
+import org.slf4j.LoggerFactory
+import java.sql.Connection
+import java.sql.SQLException
+import java.util.UUID
+
+private val logger = LoggerFactory.getLogger(PGSessionAdvisoryBookmarkLock::class.java)
+class PGSessionAdvisoryBookmarkLock(
+    private val lockTimeoutMs: Int = 60_000,
+    private val connectionProvider: () -> Connection,
```

Comment:
Will this be backed by a pool? Or so we need to somehow indicate that it's important this it's a single long-lived connection?

Reply:
In our case it will be backed by a pool. Does it matter?

Reply:
I think we figured out it matters didn't we? Because when the projector is doing it's work, it's pulling a random connection from the pool. If it did it's previous work on another connection, then the lock is living on that connection, and will just live there unused while the projector is blocked up on another connection

Reply:
This is a single connection specifically for the locking across all projectors - it's not used by any other code because we never return it to the pool (by closing it). But it is consuming a connection out of the pool that would otherwise be actively used by projectors, so theoretically it could cause performance issues. I originally created a new pool with max 1 connection for this but removed it when I encountered an error. I think the error was unrelated though so I could probably put it back

Reply:
Oh ok - when I read the code it looked like it was going to get a new connection every time - can you please show me where we're getting a connection only one from the pool and then holding on to it? I would have expected it to end up as a property on the PGSessionAdvisoryBookmarkLock

Reply:
right here: https://github.com/cultureamp/kestrel/pull/62/files/ddb646b35debbb099445c1fc3032f99267b8385b#diff-8fe24a5dc3b2a10915166c71bee3a26ae20a8f3cea7ced1505e431d9326315dcR14